<h1>Advent of Code - Day 2</h1>

<h3>Day 2 - Part 1 - Problem</h3>

<p>You get inside and take the elevator to its only other stop: the gift shop. "Thank you for visiting the North Pole!" gleefully exclaims a nearby sign. You aren't sure who is even allowed to visit the North Pole, but you know you can access the lobby through here, and from there you can access the rest of the North Pole base.

As you make your way through the surprisingly extensive selection, one of the clerks recognizes you and asks for your help.

As it turns out, one of the younger Elves was playing on a gift shop computer and managed to add a whole bunch of invalid product IDs to their gift shop database! Surely, it would be no trouble for you to identify the invalid product IDs for them, right?

They've even checked most of the product ID ranges already; they only have a few product ID ranges (your puzzle input) that you'll need to check. For example:

11-22,95-115,998-1012,1188511880-1188511890,222220-222224,
1698522-1698528,446443-446449,38593856-38593862,565653-565659,
824824821-824824827,2121212118-2121212124
(The ID ranges are wrapped here for legibility; in your input, they appear on a single long line.)

The ranges are separated by commas (,); each range gives its first ID and last ID separated by a dash (-).

Since the young Elf was just doing silly patterns, you can find the invalid IDs by looking for any ID which is made only of some sequence of digits repeated twice. So, 55 (5 twice), 6464 (64 twice), and 123123 (123 twice) would all be invalid IDs.

None of the numbers have leading zeroes; 0101 isn't an ID at all. (101 is a valid ID that you would ignore.)

Your job is to find all of the invalid IDs that appear in the given ranges. In the above example:
<ul>
    <li>11-22 has two invalid IDs, 11 and 22.</li>
    <li>95-115 has one invalid ID, 99.</li>
    <li>998-1012 has one invalid ID, 1010.</li>
    <li>1188511880-1188511890 has one invalid ID, 1188511885.</li>
    <li>222220-222224 has one invalid ID, 222222.</li>
    <li>1698522-1698528 contains no invalid IDs.</li>
    <li>446443-446449 has one invalid ID, 446446.</li>
    <li>38593856-38593862 has one invalid ID, 38593859.</li>
    <li>The rest of the ranges contain no invalid IDs.</li>

</ul>
Adding up all the invalid IDs in this example produces 1227775554.

What do you get if you add up all of the invalid IDs?
</p>

<h3>Day 2 - Part 1 - Answer</h3>

<p>Now this is a simple one, we just look at all of the numbers between the ranges, split them at the half and check if they match.</p>

<p> Let's just do a simple check to split the numbers at exactly half.</p>

In [10]:
test = '123456' 
test2 = '11112222'

first = test[0:int(len(test)/2)]
print(first)
last = test[int(len(test)/2):len(test)] 
print(last)

first = test2[0:int(len(test2)/2)]
print(first)
last = test2[int(len(test2)/2):len(test2)] 
print(last)

123
456
1111
2222


<p>That approach should work for any kind of length. Let's check some shorter ones just in case:</p>

In [13]:
test = '12' 
test2 = '1'

first = test[0:int(len(test)/2)]
print(first)
last = test[int(len(test)/2):len(test)] 
print(last)
print(first == last) 

first = test2[0:int(len(test2)/2)]
print(first)
last = test2[int(len(test2)/2):len(test2)] 
print(last)
print(first == last) 

1
2
False

1
False


<p> This approach does not seem to throw exceptions even with smaller numbers, what about odd ones?</p>

In [14]:
test = '12345' 
test2 = '1112222'

first = test[0:int(len(test)/2)]
print(first)
last = test[int(len(test)/2):len(test)] 
print(last)

first = test2[0:int(len(test2)/2)]
print(first)
last = test2[int(len(test2)/2):len(test2)] 
print(last)

12
345
111
2222


<p>This method should work, let's start:</p>

In [94]:
#Let's open our file
f = open("day2input25.txt") 
#Turn our input into an array using the split()-method
for x in f:
    ranges = x.split(",")

In [95]:
print(ranges)

['990244-1009337', '5518069-5608946', '34273134-34397466', '3636295061-3636388848', '8613701-8663602', '573252-688417', '472288-533253', '960590-988421', '7373678538-7373794411', '178-266', '63577667-63679502', '70-132', '487-1146', '666631751-666711926', '5896-10827', '30288-52204', '21847924-21889141', '69684057-69706531', '97142181-97271487', '538561-555085', '286637-467444', '93452333-93519874', '69247-119122', '8955190262-8955353747', '883317-948391', '8282803943-8282844514', '214125-236989', '2518-4693', '586540593-586645823', '137643-211684', '33-47', '16210-28409', '748488-837584', '1381-2281', '1-19\n']


<p> Now for the actual part:</p>

In [56]:
#Initializing our password
password = 0
#Iterating through our ranges
for x in ranges:
    #Splitting the range from "-"
    range1, range2 = x.split("-")
    #Checking all values between the range, edges included
    for i in range(int(range1), int(range2)+1): #end range is not inclusive so +1
        #Converting our values to string in order to split
        string = str(i) 
        #Performing the splits
        first = string[0:int(len(string)/2)]
        second = string[int(len(string)/2):len(string)] 
        #Checking if the id is invalid
        if first == second:
            #Adding invalid id to the password
            password += int(string)
            
print(password)

99


<p>And that is our correct answer!</p>

<h3>Day 2 - Part 2 - Question</h3>

<p>The clerk quickly discovers that there are still invalid IDs in the ranges in your list. Maybe the young Elf was doing other silly patterns as well?

Now, an ID is invalid if it is made only of some sequence of digits repeated at least twice. So, 12341234 (1234 two times), 123123123 (123 three times), 1212121212 (12 five times), and 1111111 (1 seven times) are all invalid IDs.

From the same example as before:

<ul>
    <li>11-22 still has two invalid IDs, 11 and 22.</li>
    <li>95-115 now has two invalid IDs, 99 and 111.</li>
    <li>998-1012 now has two invalid IDs, 999 and 1010.</li>
    <li>1188511880-1188511890 still has one invalid ID, 1188511885.</li>
    <li>222220-222224 still has one invalid ID, 222222.</li>
    <li>1698522-1698528 still contains no invalid IDs.</li>
    <li>446443-446449 still has one invalid ID, 446446.</li>
    <li>38593856-38593862 still has one invalid ID, 38593859.</li>
    <li>565653-565659 now has one invalid ID, 565656.</li>
    <li>824824821-824824827 now has one invalid ID, 824824824.</li>
    <li>2121212118-2121212124 now has one invalid ID, 2121212121.</li>
</ul>

Adding up all the invalid IDs in this example produces 4174379265.

What do you get if you add up all of the invalid IDs using these new rules?
</p>

<h3>Day 2 - Part 2 - Answer</h3>
<p>Now we have to perform multiple splices depending on the string length. So we have to check whether a pattern repeats up to half of the strings length.
</p>

In [99]:
#Initializing our password
password = 0
#Iterating through our ranges
for x in ranges:
    #Splitting the range from "-"
    range1, range2 = x.split("-")
    #Checking all values between the range, edges included
    for i in range(int(range1), int(range2)+1): #end range is not inclusive so +1
        #Converting our values to string in order to split
        string = str(i)
        #Checking all possible substring lengths.
        for k in range(1, int(len(string)//2)+1):
            #Checking if the string can be divided by the range
            if len(string)%k == 0:
                #Checking if the string can be expressed as a multiplication of the substing
                if string[0:k]*(len(string)//k) == string:
                    #Adding our correct answer and breaking the loop
                    password += int(string) 
                    break
            
print(password)

44143124633


<p> And that's the correct answer, while Googling I found an answer that struck a little too close home and tried to recreate it by memory but after 30 minutes of fiddling I decided to give in. Got to be more careful next  time.</p>